File: src/test/java/org/organicdesign/fp/xform/TransformableTest.java
package org.organicdesign.fp.xform;

import java.util.*;

import org.junit.Test;
import org.organicdesign.fp.collections.ImMap;
import org.organicdesign.fp.collections.ImSortedMap;
import org.organicdesign.fp.collections.PersistentHashMap;
import org.organicdesign.fp.collections.PersistentTreeMap;
import org.organicdesign.fp.collections.UnmodSortedIterable;
import org.organicdesign.fp.function.Fn1;
import org.organicdesign.fp.oneOf.Option;
import org.organicdesign.fp.tuple.Tuple2;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.organicdesign.fp.FunctionUtils.ordinal;

public class TransformableTest {
    static final List<Integer> controlList = Collections.unmodifiableList(Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9));

    @Test public void testAny() {
        assertTrue(Xform.of(controlList)
                        .any(i -> i > 3));
        assertTrue(Xform.of(controlList)
                        .any(i -> i > 8));
        assertTrue(Xform.of(controlList)
                        .any(i -> i < 2));
        assertTrue(Xform.of(controlList)
                        .any(i -> i == 5));
        assertFalse(Xform.of(controlList)
                        .any(i -> i == 0));
        assertFalse(Xform.of(controlList)
                         .any(Fn1.reject()));
        assertFalse(Xform.of(Collections.emptyList())
                         .any(Fn1.accept()));
        assertTrue(Xform.of(Collections.singletonList(1))
                         .any(Fn1.accept()));
    }

    @Test public void testHead() {
        // TODO: I'm not getting test coverage from this.  It seems to be using the UnmodIterable implementation instead (which is probably more efficient)
        Transformable<Integer> tint1 = Xform.of(controlList);
        assertEquals(Option.some(1), tint1.head());

        Transformable<Integer> tint2 = Xform.of(Collections.emptyList());
        assertEquals(Option.none(), tint2.head());
    }

    @Test public void testToMutList() {
        Transformable<Integer> trans = Xform.of(controlList);
        assertEquals(controlList, trans.toMutList());
    }

    @Test public void testToImList() {
        Transformable<Integer> trans = Xform.of(controlList);
        assertEquals(controlList, trans.toImList());
    }

    @Test public void testToMutRrbt() {
        Transformable<Integer> trans = Xform.of(controlList);
        assertEquals(controlList, trans.toMutRrbt());
    }

    @Test public void testToImRrbt() {
        Transformable<Integer> trans = Xform.of(controlList);
        assertEquals(controlList, trans.toImRrbt());
    }

    @Test public void testToMutMap() {
        Map<Integer,String> control = new HashMap<>();
        for (int i = 1; i < 12; i++) {
            control.put(i, ordinal(i));
        }
        Transformable<Map.Entry<Integer,String>> trans = Xform.of(control.entrySet());
        assertEquals(control, trans.toMutMap(x -> x));
    }

    @Test public void testToMutSortedMap() {
        Comparator<Integer> comp = (a, b) -> a - b;
        SortedMap<Integer,String> control = new TreeMap<>(comp);
        for (int i : controlList) {
            control.put(i, ordinal(i));
        }
        Transformable<Integer> trans = Xform.of(controlList);
        assertEquals(control, trans.toMutSortedMap(comp, i -> Tuple2.of(i, ordinal(i))));
    }

    @Test public void testToImMap() {
        ImMap<Integer,String> control = PersistentHashMap.empty();
        for (int i = 1; i < 12; i++) {
            control = control.assoc(i, ordinal(i));
        }
        Transformable<Map.Entry<Integer,String>> trans = Xform.of(control);
        assertEquals(control, trans.toImMap(x -> x));

    }

    @Test public void testToImSortedMap() {
        Comparator<Integer> comp = (a, b) -> a - b;
        ImSortedMap<Integer,String> control = PersistentTreeMap.empty(comp);
        for (int i : controlList) {
            control = control.assoc(i, ordinal(i));
        }
        Transformable<Integer> trans = Xform.of(controlList);
        assertEquals(control, trans.toImSortedMap(comp, i -> Tuple2.of(i, ordinal(i))));
    }

    @Test public void testToMutSortedSet() {
        SortedSet<Integer> control = new TreeSet<>(controlList);
        Transformable<Integer> trans = Xform.of(control);
        assertTrue(UnmodSortedIterable.equal(UnmodSortedIterable.castFromSortedSet(control),
                                             UnmodSortedIterable.castFromSortedSet(trans.toMutSortedSet((a, b) -> a - b))));
    }

    @Test public void testToImSet() {
        Set<Integer> control = new HashSet<>(controlList);
        Transformable<Integer> trans = Xform.of(controlList);
        assertEquals(control, trans.toImSet());
    }

    @Test public void testToImSortedSet() {
        Comparator<Integer> comp = (a, b) -> a - b;
        SortedSet<Integer> control = new TreeSet<>(comp);
        control.addAll(controlList);
        Transformable<Integer> trans = Xform.of(controlList);
        assertEquals(control, trans.toImSortedSet(comp));
        assertTrue(UnmodSortedIterable.equal(UnmodSortedIterable.castFromSortedSet(control),
                                             trans.toImSortedSet(comp)));
    }

    @Test public void testToMutSet() {
        Set<Integer> control = new HashSet<>(controlList);
        Transformable<Integer> trans = Xform.of(control);
        assertEquals(control, trans.toMutSet());
    }
}

File: src/test/java/org/organicdesign/fp/xform/TransformableTest.kt
// Copyright 2015 PlanBase Inc. & Glen Peterson
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.organicdesign.fp.xform

import java.util.*

import org.junit.Test
import org.organicdesign.fp.collections.ImMap
import org.organicdesign.fp.collections.ImSortedMap
import org.organicdesign.fp.collections.PersistentHashMap
import org.organicdesign.fp.collections.PersistentTreeMap
import org.organicdesign.fp.collections.UnmodSortedIterable
import org.organicdesign.fp.oneOf.Option
import org.organicdesign.fp.tuple.Tuple2

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.organicdesign.fp.FunctionUtils.ordinal
import kotlin.collections.Map.Entry

class TransformableTest {

    @Test
    @Throws(Exception::class)
    fun testToMutableList() {
        val control = Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9)
        val trans = Xform.of(control)
        assertEquals(control, trans.toMutList())
    }

    @Test
    @Throws(Exception::class)
    fun testToImList() {
        val control = Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9)
        val trans = Xform.of(control)
        assertEquals(control, trans.toImList())
    }

    @Test
    @Throws(Exception::class)
    fun testToMutableMap() {
        val control = HashMap<Int, String>()
        for (i in 1..11) {
            control.put(i, ordinal(i))
        }
        val trans = Xform.of<Entry<Int, String>>(control.entries)
        assertEquals(control, trans.toMutMap<Int, String> { x -> x })
    }

    @Test
    @Throws(Exception::class)
    fun testToMutableSortedMap() {
        val items:List<Int> = Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9)
        val comp = Integer::compare
        val control = TreeMap<Int, String>(comp)
        for (i in items) {
            control.put(i, ordinal(i))
        }
        val trans = Xform.of(items)
        assertEquals(control, trans.toMutSortedMap(comp, { i -> Tuple2.of(i, ordinal(i)) }))
    }

    @Test
    @Throws(Exception::class)
    fun testToImMap() {
        var control: ImMap<Int, String> = PersistentHashMap.empty()
        for (i in 1..11) {
            control = control.assoc(i, ordinal(i))
        }
        val trans = Xform.of<Entry<Int, String>>(control)
        assertEquals(control, trans.toImMap<Int, String> { x -> x })

    }

    @Test
    @Throws(Exception::class)
    fun testToImSortedMap() {
        val items = Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9)
        val comp = Integer::compare
        var control: ImSortedMap<Int, String> = PersistentTreeMap.empty(comp)
        for (i in items) {
            control = control.assoc(i, ordinal(i))
        }
        val trans = Xform.of(items)
        assertEquals(control, trans.toImSortedMap(comp) { i -> Tuple2.of(i, ordinal(i!!)) })
    }

    @Test
    @Throws(Exception::class)
    fun testToMutableSortedSet() {
        val control = TreeSet(Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9))
        val trans = Xform.of(control)
        assertTrue(UnmodSortedIterable.equal(UnmodSortedIterable.castFromSortedSet(control),
                                             UnmodSortedIterable.castFromSortedSet(trans.toMutSortedSet { a, b -> a!! - b!! })))
    }

    @Test
    @Throws(Exception::class)
    fun testToImSet() {
        val items = Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9)
        val control = HashSet(items)
        val trans = Xform.of(items)
        assertEquals(control, trans.toImSet())
    }

    @Test
    @Throws(Exception::class)
    fun testToImSortedSet() {
        val items = Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9)
        val comp = Integer::compare
        val control = TreeSet<Int>(comp)
        control.addAll(items)
        val trans = Xform.of(items)
        assertEquals(control, trans.toImSortedSet(comp))
        assertTrue(UnmodSortedIterable.equal(UnmodSortedIterable.castFromSortedSet(control),
                                             trans.toImSortedSet(comp)))
    }

    @Test
    @Throws(Exception::class)
    fun testToMutableSet() {
        val control = HashSet(Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9))
        val trans = Xform.of(control)
        assertEquals(control, trans.toMutSet())
    }

    @Test
    fun testHead() {
        // TODO: I'm not getting test coverage from this.  It seems to be using the UnmodIterable implementation instead (which is probably more efficient)
        val tint1 = Xform.of(Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9))
        assertEquals(Option.some(1), (tint1 as Transformable<Int>).head())

        val tint2 = Xform.of(emptyList<Int>())
        assertEquals(Option.none<Any>(), (tint2 as Transformable<Int>).head())
    }
}